community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/idp/IDPQueryGraphSolver.scala

```diff
@@ -44,6 +44,76 @@ object IDPQueryGraphSolver {
   val VERBOSE = java.lang.Boolean.getBoolean("pickBestPlan.VERBOSE")
 }
 
+/* The IDP inner loop can be optimized and tweaked in several ways, and this trait encapsulates those settings */
+trait IDPSolverConfig {
+  def maxTableSize: Int
+  def iterationDurationLimit(queryGraph: QueryGraph): Long
+  def solvers(queryGraph: QueryGraph): Seq[QueryGraph => IDPSolverStep[PatternRelationship, LogicalPlan, LogicalPlanningContext]]
+}
+
+/* The Dynamic Programming (DP) approach is IDP with no optimizations */
+case class DPSolverConfig(maxTableSize: Int = Integer.MAX_VALUE, iterationDuration: Long = Long.MaxValue) extends IDPSolverConfig {
+  override def iterationDurationLimit(queryGraph: QueryGraph) = iterationDuration
+  override def solvers(queryGraph: QueryGraph) = Seq(joinSolverStep(_), expandSolverStep(_))
+}
+
+/* The default settings for IDP uses a maxTableSize and a inner loop duration threshold
+   to improve planning performance with minimal impact of plan quality */
+case class DefaultIDPSolverConfig(maxTableSize: Int = 128, iterationDuration: Long = 5000) extends IDPSolverConfig {
+  override def iterationDurationLimit(queryGraph: QueryGraph) = iterationDuration
+  override def solvers(queryGraph: QueryGraph) = Seq(joinSolverStep(_), expandSolverStep(_))
+}
+
+/* For testing IDP we sometimes limit the solver to expands only */
+case class ExpandOnlyIDPSolverConfig(maxTableSize: Int = 128, iterationDuration: Long = 5000) extends IDPSolverConfig {
+  override def iterationDurationLimit(queryGraph: QueryGraph) = iterationDuration
+  override def solvers(queryGraph: QueryGraph) = Seq(expandSolverStep(_))
+}
+
+/* For testing IDP we sometimes limit the solver to joins only */
+case class JoinOnlyIDPSolverConfig(maxTableSize: Int = 128, iterationDuration: Long = 5000) extends IDPSolverConfig {
+  override def iterationDurationLimit(queryGraph: QueryGraph) = iterationDuration
+  override def solvers(queryGraph: QueryGraph) = Seq(joinSolverStep(_))
+}
+
+/* One approach to optimizing planning of very long patterns is to only allow expands.
+   This making planning very fast, but can lead to non-optimal plans where joins make more sense */
+case class ExpandOnlyWhenPatternIsLong(maxTableSize: Int = 128, iterationDuration: Long = 1000) extends IDPSolverConfig {
+  override def iterationDurationLimit(queryGraph: QueryGraph) = iterationDuration
+  override def solvers(queryGraph: QueryGraph) =
+    if(queryGraph.patternRelationships.size > 10) Seq(expandSolverStep(_))
+    else Seq(joinSolverStep(_), expandSolverStep(_))
+}
+
+/* Another approach to optimizing planning of very long patterns is to reduce the inner loop time threshold, making IDP less exhaustive.
+   This making planning very fast, but can lead to non-optimal plans due to local minima. */
+case class ReducingTimeBoxPatternConfig(maxTableSize: Int = 128, patternLengthThreshold: Int = 10) extends IDPSolverConfig {
```

Comment:
The intention was to allow calling code (include test code) to change the settings, which means they are not singletons. I suggest I clarify that by removing the default parameters instead.